sql/expression/function/aggregation/json_agg.go

```diff
@@ -32,21 +36,102 @@ var ErrUnsupportedJSONFunction = errors.NewKind("unsupported JSON function: %s")
 //
 // see also: https://dev.mysql.com/doc/refman/8.0/en/json.html#json-normalization
 type JSONArrayAgg struct {
-	sql.Expression
+	expression.UnaryExpression
 }
 
-var _ sql.FunctionExpression = JSONArrayAgg{}
+var _ sql.FunctionExpression = &JSONArrayAgg{}
 
 // NewJSONArrayAgg creates a new JSONArrayAgg function.
-func NewJSONArrayAgg(args ...sql.Expression) (sql.Expression, error) {
-	return nil, ErrUnsupportedJSONFunction.New(JSONArrayAgg{}.FunctionName())
+func NewJSONArrayAgg(arg sql.Expression) *JSONArrayAgg {
+	return &JSONArrayAgg{expression.UnaryExpression{Child: arg}}
 }
 
 // FunctionName implements sql.FunctionExpression
-func (j JSONArrayAgg) FunctionName() string {
+func (j *JSONArrayAgg) FunctionName() string {
 	return "json_arrayagg"
 }
 
+// NewBuffer creates a new buffer for the aggregation.
+func (j *JSONArrayAgg) NewBuffer() sql.Row {
+	var row []interface{}
+	return sql.NewRow(row)
+}
+
+// Type returns the type of the result.
+func (j *JSONArrayAgg) Type() sql.Type {
+	return sql.JSON
+}
+
+// IsNullable returns whether the return value can be null.
+func (j *JSONArrayAgg) IsNullable() bool {
+	return false
+}
+
+// Resolved implements the Expression interface.
+func (j *JSONArrayAgg) Resolved() bool {
+	if _, ok := j.Child.(*expression.Star); ok {
+		return true
+	}
+
+	return j.Child.Resolved()
+}
+
+func (j *JSONArrayAgg) String() string {
+	return fmt.Sprintf("JSON_ARRAYAGG(%s)", j.Child)
+}
+
+// WithChildren implements the Expression interface.
+func (j *JSONArrayAgg) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 1 {
+		return nil, sql.ErrInvalidChildrenNumber.New(j, len(children), 1)
+	}
+	return NewJSONArrayAgg(children[0]), nil
+}
+
+// Update implements the Aggregation interface.
+func (j *JSONArrayAgg) Update(ctx *sql.Context, buffer, row sql.Row) error {
+	v, err := j.Child.Eval(ctx, row)
+	if err != nil {
+		return err
+	}
+
+	buffer[0] = append(buffer[0].([]interface{}), v)
+
+	return nil
+}
+
+// Merge implements the Aggregation interface.
+func (j *JSONArrayAgg) Merge(ctx *sql.Context, buffer, partial sql.Row) error {
+	arr1 := buffer[0].([]interface{})
+	arr2 := partial[0].([]interface{})
+
+	buffer[0] = append(arr1, arr2...)
+
+	return nil
+}
+
+// Eval implements the Aggregation interface.
+func (j *JSONArrayAgg) Eval(ctx *sql.Context, buffer sql.Row) (interface{}, error) {
```

Comment:
as we discussed on chat, we're going to move to an in-memory representation of JSON that is in line with Go:
https://golang.org/pkg/encoding/json/#Unmarshal

Reply:
yup. Going to go ahead and merge for now.